hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/protocol/DatanodeDetails.java

```diff
@@ -175,6 +178,34 @@ public void setIpAddress(String ip) {
     this.ipAddress = StringWithByteString.valueOf(ip);
   }
 
+  /**
+   * Resolves and validates the IP address of the datanode based on its hostname.
+   * If the resolved IP address differs from the current IP address,
+   * it updates the IP address to the newly resolved value.
+   */
+  public void validateDatanodeIpAddress() {
+    final String oldIP = getIpAddress();
```

Comment:
```
NullPointerException: Cannot invoke "org.apache.hadoop.ozone.util.StringWithByteString.getString()" because "this.ipAddress" is null
```

Sorry, it looks like we need to handle `ipAddress == null` case in `getIpAddress()` and `getHostName()`.